arch/x86/x86.c

```diff
@@ -14,6 +14,22 @@ ZLIB_INTERNAL int x86_cpu_has_sse2;
 ZLIB_INTERNAL int x86_cpu_has_sse42;
 ZLIB_INTERNAL int x86_cpu_has_pclmulqdq;
 
+#ifdef _MSC_VER
+#include <intrin.h>
+#define CPU_PROCINFO_AND_FEATUREBITS 1
+
+void ZLIB_INTERNAL x86_check_features(void) {
+	unsigned int registers[4];
+	__cpuid(registers, CPU_PROCINFO_AND_FEATUREBITS);
+
+	unsigned int ecx = registers[2];
+	unsigned int edx = registers[3];
+
+	x86_cpu_has_sse2 = edx & 0x4000000;
```

Comment:
For future expansion of the cpuid capability when we need to check for more settings, it would be nice if this is as obvious as possible.

But I am thinking that it might be better to only duplicate the cpuid instruction and population of the variables, while the actual checking of features should not be duplicated.

In other words, the following part should be common code, while the code to populate eax,ebx,ecx and edx variables should be split out to a new function (x86_get_cpuid for example) selected by defines. This way both parts can be easily maintained, readability is increased and addition of checks for more cpuid features can be implemented with minimal chances of breaking the compiler-dependant parts of the code.
```
     x86_cpu_has_sse2 = edx & 0x4000000;
     x86_cpu_has_sse42 = ecx & 0x100000;
     x86_cpu_has_pclmulqdq = ecx & 0x2;
```